Removing a secondary control plane load balancer panics the AWSCluster validating webhook

/kind bug

**What steps did you take and what happened:**

Updating an `AWSCluster` to remove a previously configured `spec.secondaryControlPlaneLoadBalancer` by setting the field to `null` causes the validating webhook handler to panic with a nil pointer dereference:

```text
panic: runtime error: invalid memory address or nil pointer dereference
...validateControlPlaneLoadBalancerUpdate(..., <non-nil oldlb>, 0x0)
    webhooks/awscluster_webhook.go:205
...ValidateUpdate(...)
    webhooks/awscluster_webhook.go:128
```

`ValidateUpdate` constructs old/new pairs for the primary and secondary control plane load balancer specifications. It skips validation only when both values are nil.

When an existing secondary load balancer configuration is removed:

```text
oldlb != nil
newlb == nil
```

the pair is passed to `validateControlPlaneLoadBalancerUpdate`. The function enters the `oldlb != nil` branch and dereferences `newlb`, for example through `newlb.LoadBalancerType`, without first checking whether `newlb` is nil.

The Go HTTP server recovers the panic for the individual request, so the CAPA manager process remains running. However, the admission request is aborted and the `AWSCluster` update fails with an internal webhook error rather than a normal validation response.

This is the mirror case of #5247, fixed by #5248:

* #5247 handled adding a secondary load balancer: `oldlb == nil`, `newlb != nil`.
* This issue concerns removing one: `oldlb != nil`, `newlb == nil`.

The removal case is not currently covered by the update validation tests.

**What did you expect to happen:**

The validating webhook must not panic for this update.

If removing an existing secondary control plane load balancer is supported, the update should be handled safely and the corresponding AWS resources should be reconciled appropriately.

If removal is not supported, the webhook should reject the update with a clear validation error explaining that the field cannot be removed.

**Anything else you would like to add:**

This can be reproduced at the webhook validation level by calling `ValidateUpdate` with:

* an old `AWSCluster` containing a non-nil `SecondaryControlPlaneLoadBalancer`; and
* a new `AWSCluster` where `SecondaryControlPlaneLoadBalancer` is nil.

A regression test should verify both that the call does not panic and that it returns the intended result: either successful validation or an explicit validation error.

I would be happy to open a PR once the intended removal semantics are confirmed. A minimal fix could explicitly reject removal without panicking; fully supporting removal may additionally require reconciliation and cleanup of the existing AWS load balancer resources.

**Environment:**

* Cluster API Provider AWS version: current `main`
* Kubernetes version: n/a; this occurs in AWSCluster webhook update validation

--- a/webhooks/awscluster_webhook.go
+++ b/webhooks/awscluster_webhook.go
@@ -114,18 +114,24 @@ func (w *AWSCluster) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Ob
 		)
 	}
 
-	// Validate the control plane load balancers.
-	lbs := map[*infrav1.AWSLoadBalancerSpec]*infrav1.AWSLoadBalancerSpec{
-		oldC.Spec.ControlPlaneLoadBalancer:          r.Spec.ControlPlaneLoadBalancer,
-		oldC.Spec.SecondaryControlPlaneLoadBalancer: r.Spec.SecondaryControlPlaneLoadBalancer,
+	// Validate the control plane load balancers. Each entry carries its own field
+	// path so validation errors (including the immutability checks below) are
+	// attributed to the correct load balancer instead of always to the primary.
+	lbs := []struct {
+		old  *infrav1.AWSLoadBalancerSpec
+		new  *infrav1.AWSLoadBalancerSpec
+		path *field.Path
+	}{
+		{oldC.Spec.ControlPlaneLoadBalancer, r.Spec.ControlPlaneLoadBalancer, field.NewPath("spec", "controlPlaneLoadBalancer")},
+		{oldC.Spec.SecondaryControlPlaneLoadBalancer, r.Spec.SecondaryControlPlaneLoadBalancer, field.NewPath("spec", "secondaryControlPlaneLoadBalancer")},
 	}
 
-	for oldLB, newLB := range lbs {
-		if oldLB == nil && newLB == nil {
+	for _, lb := range lbs {
+		if lb.old == nil && lb.new == nil {
 			continue
 		}
 
-		allErrs = append(allErrs, w.validateControlPlaneLoadBalancerUpdate(oldLB, newLB)...)
+		allErrs = append(allErrs, w.validateControlPlaneLoadBalancerUpdate(lb.old, lb.new, lb.path)...)
 	}
 
 	if !cmp.Equal(oldC.Spec.ControlPlaneEndpoint, clusterv1beta1.APIEndpoint{}) &&
@@ -186,14 +192,25 @@ func (w *AWSCluster) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Ob
 	return allWarnings, aggregateObjErrors(r.GroupVersionKind().GroupKind(), r.Name, allErrs)
 }
 
-func (w *AWSCluster) validateControlPlaneLoadBalancerUpdate(oldlb, newlb *infrav1.AWSLoadBalancerSpec) field.ErrorList {
+func (w *AWSCluster) validateControlPlaneLoadBalancerUpdate(oldlb, newlb *infrav1.AWSLoadBalancerSpec, lbPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
+
+	// A previously-set control plane load balancer cannot be removed: the
+	// underlying AWS load balancer is not deleted by reconciliation, so allowing
+	// the removal would leak it. Reject the change with a clear, field-scoped
+	// error instead of dereferencing newlb below, which would panic when it is
+	// nil.
+	if oldlb != nil && newlb == nil {
+		allErrs = append(allErrs, field.Invalid(lbPath, newlb,
+			"control plane load balancer cannot be removed once set"))
+		return allErrs
+	}
 
 	if oldlb == nil {
 		// If old scheme was nil, the only value accepted here is the default value: internet-facing
 		if newlb.Scheme != nil && newlb.Scheme.String() != infrav1.ELBSchemeInternetFacing.String() {
 			allErrs = append(allErrs,
-				field.Invalid(field.NewPath("spec", "controlPlaneLoadBalancer", "scheme"),
+				field.Invalid(lbPath.Child("scheme"),
 					newlb.Scheme, "field is immutable, default value was set to internet-facing"),
 			)
 		}
@@ -204,14 +221,14 @@ func (w *AWSCluster) validateControlPlaneLoadBalancerUpdate(oldlb, newlb *infrav
 		if (oldlb.LoadBalancerType == infrav1.LoadBalancerTypeDisabled && newlb.LoadBalancerType != infrav1.LoadBalancerTypeDisabled) ||
 			(newlb.LoadBalancerType == infrav1.LoadBalancerTypeDisabled && oldlb.LoadBalancerType != infrav1.LoadBalancerTypeDisabled) {
 			allErrs = append(allErrs,
-				field.Invalid(field.NewPath("spec", "controlPlaneLoadBalancer", "type"),
+				field.Invalid(lbPath.Child("type"),
 					newlb.Scheme, "field is immutable when created of disabled type"),
 			)
 		}
 		// If old scheme was not nil, the new scheme should be the same.
 		if !cmp.Equal(oldlb.Scheme, newlb.Scheme) {
 			allErrs = append(allErrs,
-				field.Invalid(field.NewPath("spec", "controlPlaneLoadBalancer", "scheme"),
+				field.Invalid(lbPath.Child("scheme"),
 					newlb.Scheme, "field is immutable"),
 			)
 		}
@@ -220,7 +237,7 @@ func (w *AWSCluster) validateControlPlaneLoadBalancerUpdate(oldlb, newlb *infrav
 		// so the name remains nil. In either case, the name cannot be changed.
 		if !cmp.Equal(oldlb.Name, newlb.Name) {
 			allErrs = append(allErrs,
-				field.Invalid(field.NewPath("spec", "controlPlaneLoadBalancer", "name"),
+				field.Invalid(lbPath.Child("name"),
 					newlb.Name, "field is immutable"),
 			)
 		}
@@ -231,7 +248,7 @@ func (w *AWSCluster) validateControlPlaneLoadBalancerUpdate(oldlb, newlb *infrav
 		if oldlb.LoadBalancerType != infrav1.LoadBalancerTypeClassic {
 			if !cmp.Equal(newlb.HealthCheckProtocol, oldlb.HealthCheckProtocol) {
 				allErrs = append(allErrs,
-					field.Invalid(field.NewPath("spec", "controlPlaneLoadBalancer", "healthCheckProtocol"),
+					field.Invalid(lbPath.Child("healthCheckProtocol"),
 						newlb.HealthCheckProtocol, "field is immutable once set"),
 				)
 			}
@@ -240,7 +257,7 @@ func (w *AWSCluster) validateControlPlaneLoadBalancerUpdate(oldlb, newlb *infrav
 		// TargetGroupIPType is immutable after creation.
 		if !cmp.Equal(oldlb.TargetGroupIPType, newlb.TargetGroupIPType) {
 			allErrs = append(allErrs,
-				field.Forbidden(field.NewPath("spec", "controlPlaneLoadBalancer", "targetGroupIPType"),
+				field.Forbidden(lbPath.Child("targetGroupIPType"),
 					"field is immutable and cannot be changed after target group creation"),
 			)
 		}

--- a/webhooks/awscluster_webhook_test.go
+++ b/webhooks/awscluster_webhook_test.go
@@ -25,6 +25,7 @@ import (
 
 	"github.com/aws/aws-sdk-go-v2/aws"
 	. "github.com/onsi/gomega"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	utilfeature "k8s.io/component-base/featuregate/testing"
@@ -1519,6 +1520,31 @@ func TestAWSClusterValidateUpdate(t *testing.T) {
 			wantErr: false,
 		},
 		{
+			name: "Should fail if the secondary control plane load balancer is removed",
+			oldCluster: &infrav1.AWSCluster{
+				Spec: infrav1.AWSClusterSpec{
+					ControlPlaneLoadBalancer: &infrav1.AWSLoadBalancerSpec{
+						Name:   ptr.To("primary-lb"),
+						Scheme: &infrav1.ELBSchemeInternetFacing,
+					},
+					SecondaryControlPlaneLoadBalancer: &infrav1.AWSLoadBalancerSpec{
+						Name:             ptr.To("secondary-lb"),
+						Scheme:           &infrav1.ELBSchemeInternal,
+						LoadBalancerType: infrav1.LoadBalancerTypeNLB,
+					},
+				},
+			},
+			newCluster: &infrav1.AWSCluster{
+				Spec: infrav1.AWSClusterSpec{
+					ControlPlaneLoadBalancer: &infrav1.AWSLoadBalancerSpec{
+						Name:   ptr.To("primary-lb"),
+						Scheme: &infrav1.ELBSchemeInternetFacing,
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
 			name: "Should pass if controlPlaneLoadBalancer healthcheckprotocol is same after update",
 			oldCluster: &infrav1.AWSCluster{
 				Spec: infrav1.AWSClusterSpec{
@@ -1682,6 +1708,45 @@ func TestAWSClusterValidateUpdate(t *testing.T) {
 		},
 		)
 	}
+}
+
+func TestAWSClusterValidateUpdateSecondaryControlPlaneLoadBalancerRemovalIsRejected(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	primaryScheme := infrav1.ELBSchemeInternetFacing
+	secondaryScheme := infrav1.ELBSchemeInternal
+
+	oldCluster := &infrav1.AWSCluster{
+		ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "default"},
+		Spec: infrav1.AWSClusterSpec{
+			ControlPlaneLoadBalancer: &infrav1.AWSLoadBalancerSpec{
+				Name:   ptr.To("primary-lb"),
+				Scheme: &primaryScheme,
+			},
+			SecondaryControlPlaneLoadBalancer: &infrav1.AWSLoadBalancerSpec{
+				Name:             ptr.To("secondary-lb"),
+				Scheme:           &secondaryScheme,
+				LoadBalancerType: infrav1.LoadBalancerTypeNLB,
+			},
+		},
+	}
+	newCluster := oldCluster.DeepCopy()
+	newCluster.Spec.SecondaryControlPlaneLoadBalancer = nil
+
+	var err error
+	g.Expect(func() {
+		_, err = (&AWSCluster{}).ValidateUpdate(ctx, oldCluster, newCluster)
+	}).ToNot(Panic())
+	g.Expect(err).To(HaveOccurred())
+	g.Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+	statusErr, ok := err.(apierrors.APIStatus)
+	g.Expect(ok).To(BeTrue())
+	invalidFields := make([]string, 0)
+	for _, cause := range statusErr.Status().Details.Causes {
+		invalidFields = append(invalidFields, cause.Field)
+	}
+	g.Expect(invalidFields).To(ContainElement("spec.secondaryControlPlaneLoadBalancer"))
 }
 
 func TestAWSClusterDefaultCNIIngressRules(t *testing.T) {

